plugins/qlcplus_4.12.1/presets.js

```diff
@@ -0,0 +1,1002 @@
+  isBeamAngle: cap => (cap.type === `BeamAngle` || cap.type === `Zoom`) && cap.angle !== null,
+  isWheelChannel: channel => channel.capabilities.some(cap => [`WheelSlot`, `WheelRotation`].includes(cap.type)),
+  isAllowedInWheels: cap => [`WheelSlot`, `WheelShake`, `WheelSlotRotation`, `WheelRotation`, `Effect`, `NoFunction`].includes(cap.type),
+
+  /**
+   * @param {Entity} entity The speed Entity object.
+   * @returns {number|null} The frequency in Hertz, or null, if the entity's unit is not convertable to Hertz.
+   */
+  getFrequencyInHertz(entity) {
+    if (entity.unit === `Hz`) {
+      return entity.number;
+    }
+
+    if (entity.unit === `bpm`) {
+      return entity.number / 60;
+    }
+
+    return null;
+  }
+};
+
+const importHelpers = {
+  getColorIntensityCap: color => ({
+    type: `ColorIntensity`,
+    color
+  }),
+
+  getPanTiltCap: (panOrTilt, maxValue) => {
+    const cap = {
+      type: panOrTilt,
+      angleStart: `0deg`,
+      angleEnd: `${maxValue}deg`
+    };
+
+    if (maxValue === 0) {
+      cap.angleStart = `0%`;
+      cap.angleEnd = `100%`;
+      cap.helpWanted = `Can you provide exact angles?`;
+    }
+
+    return cap;
+  },
+
+  getShutterStrobeCap: (shutterEffect, speedStart, speedEnd, randomTiming) => {
+    const cap = {
+      type: `ShutterStrobe`,
+      shutterEffect
+    };
+
+    if (speedEnd) {
+      cap.speedStart = speedStart;
+      cap.speedEnd = speedEnd;
+    }
+    else if (speedStart) {
+      cap.speed = speedStart;
+    }
+
+    if (randomTiming) {
+      cap.randomTiming = true;
+    }
+
+    return cap;
+  },
+
+  getRotationCapType: ({ channelName, channelType }) => {
+    if ([`Pan`, `Tilt`].includes(channelType)) {
+      return `${channelType}Continuous`;
+    }
+
+    if ([`Colour`, `Gobo`].includes(channelType) || /\bgobo\b/i.test(channelName)) {
+      return /wheel\b/i.test(channelName) ? `WheelRotation` : `WheelSlotRotation`;
+    }
+
+    if (channelType === `Prism`) {
+      return `PrismRotation`;
+    }
+
+    return `Rotation`;
+  },
+
+  getRotationSpeedCap: (capData, speedStart, speedEnd) => {
+    const cap = {
+      type: importHelpers.getRotationCapType(capData)
+    };
+
+    if (cap.type.startsWith(`Wheel`) && !capData.channelNameInWheels) {
+      cap.wheel = ``;
+    }
+
+    if (speedEnd) {
+      cap.speedStart = speedStart;
+      cap.speedEnd = speedEnd;
+    }
+    else if (speedStart) {
+      cap.speed = speedStart;
+    }
+
+    return cap;
+  },
+
+  /**
+   * Try to guess speedStart / speedEnd from the capability name and set them
+   * to the capability. It may also set cap.type to "Rotation".
+   * @param {string} capabilityName The capability name to extract information from.
+   * @param {object} cap The OFL capability object to add found properties to.
+   * @returns {string} The rest of the capabilityName.
+   */
+  getSpeedGuessedComment(capabilityName, cap) {
```

Comment:
This function sets some properties, but to ensure a consistent property order, a lot of extra logic is needed from where it is called. Furthermore, there are some (unexpected) side effect it does, for example setting the capability type.

It may be worth refactoring this to return an object `{ type, speedStart, speedEnd, comment }` where any of the properties may be an empty string. Then set the cap data like so:

```js
const { type, speedStart, speedEnd, comment } = getSpeedData(capabilityName, cap);
cap.type = type || cap.type;
cap.effectName = comment;
if (speedStart) {
  cap.shakeSpeedStart = speedStart;
  cap.shakeSpeedEnd = speedEnd;
}
```

Reply:
Extracted to #989.